File: Javascript/3.Ottobre/70.HtmlAttributeExtractor-19Ottobre.js
//todo  Data una stringa di un elemento HTML valido, restituisci gli attributi dell'elemento utilizzando i seguenti criteri:

//! Ti verrà fornito un solo elemento.
//! Gli attributi saranno nel formato: attributo="valore".
//! Restituisci un array di stringhe con ogni proprietà e valore dell'attributo, separati da una virgola, in questo formato: ["attributo1, valore1", "attributo2, valore2"].
//! Restituisci gli attributi nell'ordine in cui sono specificati.
//! Se non vengono trovati attributi, restituisci un array vuoto.

//? Tempo di esecuzione:
//* Non completato 

//?Codice scritto senza utilizzo di ChatGpt

function extractAttributes(element) {

    let regex = /\s(\w+)=/;
    let primoValore = element.match(regex)[1]
    let regexValore = /\s(\w+)/;
    let secondoValore = element.match(regexValore)[0]

    return {primoValore, secondoValore}
}

console.log(extractAttributes('<span class="red"></span>'))

//?Codice scritto da ChatGpt

function extractAttributes(element) {
  const regex = /(\w+)="([^"]*)"/g
  const result = []
  let match

  while ((match = regex.exec(element)) !== null) {
    result.push(`${match[1]}, ${match[2]}`)
  }

  return result
}

console.log(extractAttributes('<span class="red"></span>'))

//* ✅ PRO

//* Hai capito che le regex sono la strada giusta

//* Usi match() correttamente

//* Il codice è leggibile e ordinato

//* Buona intuizione nel separare nome e valore

//! ❌ CONTRO

//! Prende solo il primo attributo (niente g)

//! regexValore non prende il valore dell’attributo, ma solo testo dopo uno spazio

//! Se non ci sono attributi → errore (match è null)

//! Non rispetta il formato richiesto (array, non oggetto)

//! Non gestisce più attributi

//! Regex troppo generiche (\w+ non copre data-*, aria-*)

//todo Considerazioni: 

//! N.B.

File: Python/3.Ottobre/70.Html_attribute_extractor-19Ottobre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# =========================================================

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

# Dato un elemento HTML valido, restituisci gli attributi
# dell'elemento come array di stringhe nel formato:
# ["attributo1, valore1", "attributo2, valore2"].
#
# Gli attributi sono sempre nel formato:
# attributo="valore"
#
# Devono essere restituiti nell'ordine originale.
# Se non sono presenti attributi, restituire un array vuoto.

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

import re

def extract_attributes(element):
    regex = r'(\w+)="([^"]*)"'
    
    matches = re.findall(regex, element)
    
    result = [f"{attr}, {value}" for attr, value in matches]
    
    return result

print(extract_attributes('<span class="red"></span>'))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

import re

def extract_attributes(element):
    return [
        f"{attribute}, {value}"
        for attribute, value in re.findall(r'(\w+)="([^"]*)"', element)
    ]

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Regex semplice e corretta.
# - Utilizzo appropriato di re.findall().
# - Comprensione di lista pulita e leggibile.
# - Mantiene l'ordine degli attributi.
# - Restituisce automaticamente una lista vuota se non trova corrispondenze.

# //! Contro
# - La variabile regex viene usata una sola volta: può essere inserita direttamente nella findall().
# - La variabile result è superflua: si può restituire direttamente la list comprehension.
# - Il pattern (\w+) non riconosce attributi con caratteri come "-" (es. data-id o aria-label), anche se la consegna potrebbe non richiederli.

# =========================================================
# //* Consigli
# =========================================================

# - Quando una variabile viene usata una sola volta, valuta se eliminarla per rendere il codice più compatto.
# - Continua ad usare le list comprehension: sono uno degli strumenti più utili di Python.
# - Abituati a chiederti se una regex copre tutti i casi possibili (ad esempio attributi con "-"), anche quando la consegna è semplice.
# - In questo esercizio la soluzione è già molto buona: è leggibile, efficiente e facile da mantenere.